Add support for ``async_simple_cache_middleware``

fselmo opened the pull request:
### What was wrong?

Related to Issue #2578

Hard-coding an internal `chain_id` was considered at some point but ultimately we landed on adding `eth_chainId` to the list of whitelisted endpoints for `simple_cache_middleware`. The `chain_id` setter had made its way into the library already and some users had gotten around the multiple `eth_chainId` checks by using this setter, especially with `AsyncEth`.

Rather than introduce this setter back, this PR adds support to cache `eth_chainId` calls using the `async_simple_cache_middleware` with `AsyncEth`.

### How was it fixed?

- Async request caching support via `async_simple_cache_middleware`
- Supports creating custom simple caching middleware via `async_construct_simple_cache_middleware`
- Refactor ``SessionCache`` to a more generic ``SimpleCache`` within the caching `_utils` and use this cache class in the simple cache middleware as the default cache (for sync and async). Leave support for a simple dictionary cache as well.

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/master/newsfragments/README.md)

#### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fwww.wallpaperflare.com%2Fstatic%2F615%2F441%2F129%2Ftitle-macro-photo-ladybug-wallpaper.jpg&f=1&nofb=1)

fselmo committed:
Add support for ``async_simple_cache_middleware``

- Add async support for simple cache middleware
- Refactor ``SessionCache`` as a more generic ``SimpleCache`` class to be used internally as a standardized cache where appropriate.
- Use ``SimpleCache`` as the default cache for the simple cache middleware

fselmo committed:
Update simple cache middleware tests

- Include tests for dictionary-based cache and ``SimpleCache`` for simple cache middleware

fselmo committed:
Clean up unnecessary test teardowns via @pacrob

kclowes reviewed (approved):
I had a couple nits, but looks good once we pull out the dictionary cache! :shipit:

fselmo committed:
Remove dictionary cache support for simple-cache-middleware

- Remove dictionary cache support altogether, for simple cache middleware, in favor of using the ``SimpleCache`` class.
- Handle missing keys in ``SimpleCache`` a bit more gracefully than throwing a ``KeyError``, just return ``None`` if the key is not in the cache.

pacrob reviewed (approved):
lgtm!

kclowes reviewed (approved):
:shipit: ! Sweet!
